docker_cli/_cli.py

```diff
@@ -0,0 +1,327 @@
+import argparse
+import sys
+import textwrap
+from typing import Sequence
+
+from ._utils import universal_tag_prefix
+from .commands import dropin, make_lockfile, remove
+from .setup_commands import (setup_all, setup_cuda_dev, setup_cuda_runtime,
+                             setup_env_add, setup_env_dev, setup_env_runtime,
+                             setup_init)
+
+
+def setup_parser() -> argparse.ArgumentParser:
+    """
+    Create a top-level argument parser.
+
+    Returns
+    -------
+    argparse.ArgumentParser
+        The parser.
+    """
+    prefix = universal_tag_prefix()
+
+    def help_formatter(prog): return argparse.HelpFormatter(
+        prog,
+        max_help_position=60
+    )
+
+    parser = argparse.ArgumentParser(
+        prog=__package__,
+        formatter_class=help_formatter
+    )
+
+    # Additional parsers for shared commands
+    setup_parse = argparse.ArgumentParser(add_help=False)
+    setup_parse.add_argument(
+        "--base", "-b", type=str, required=True,
+        help='The name of the base docker image.'
+    )
+
+    no_cache_parse = argparse.ArgumentParser(add_help=False)
+    no_cache_parse.add_argument(
+        "--no-cache",
+        action="store_true",
+        help="Run docker build with no cache if used."
+    )
+
+    cuda_run_parse = argparse.ArgumentParser(add_help=False)
+    cuda_run_parse.add_argument(
+        "--cuda-version", "-c", default="11.4", type=str,
+        help='The CUDA version. Default: "11.4"',
+        metavar="VERSION"
+    )
+    cuda_run_parse.add_argument(
+        "--cuda-repo", default="rhel8", type=str,
+        help="The name of the CUDA repository for this distro "
+        '(e.g. "rhel8", "ubuntu2004".) Default: "rhel8"',
+        metavar="REPO_NAME"
+    )
+
+    env_parse = argparse.ArgumentParser(add_help=False)
+    env_parse.add_argument(
+        "--env-file", default="spec-file.txt", type=str,
+        help="The location of the spec-file. Default: \"spec-file.txt\""
+    )
+
+    # Add arguments
+    subparsers = parser.add_subparsers(dest="command")
```

Comment:
Let's make the sub-command required so that if a user tries to invoke the top-level command without a sub-command, they'll get a helpful error message.

```suggestion
    subparsers = parser.add_subparsers(dest="command", required=True)
```

Reply:
I'm currently handling situations like this by checking in the `main()` function and then passing it along to the `insufficient_subcommands_message()` function to print a short, easily readable error. My thinking behind this is that those error messages can be a bit overly verbose and might just confuse the user about what to do next. My print statement just tells them to use `{this subcommand} -h` to get more info. Would you prefer I do the required method instead?

Reply:
I think my preference would be to make the subcommand required instead of using `insufficient_subcommands_message()` since that makes the code somewhat simpler.

If you do that and the user fails to provide a subcommand name, then the error message they get will include the help message. I agree that's more verbose than the current message, but I don't think that's a bad thing -- it's very common for command-line tools to print the help message when there's an argument parsing error, and I think many users will expect this behavior.